### PR TITLE
add span widget

### PR DESCRIPTION
**Pull request by guoweilkd**
### add span widget

create a new widget for to support rich text.

### CODE REVIEW

**Comment by guoweilkd** [comment]
example and result:
```c
LV_FONT_DECLARE(lv_font_montserrat_24)
LV_FONT_DECLARE(lv_font_montserrat_22)

static void span_example1(void)
{
    static lv_style_t style;
    lv_style_init(&style);
    lv_style_set_border_width(&style, 1);
    lv_style_set_border_color(&style, lv_palette_main(LV_PALETTE_ORANGE));
    lv_style_set_pad_all(&style, 2);

    lv_obj_t * spans = lv_spangroup_create(lv_scr_act());
    lv_obj_set_width(spans, 300);
    lv_obj_set_height(spans,300);
    lv_obj_center(spans);
    lv_obj_add_style(spans, &style, 0);

    lv_span_set_align(spans, LV_SPAN_ALIGN_LEFT);
    lv_span_set_overflow(spans, LV_SPAN_OVERFLOW_CLIP);
    lv_span_set_indent(spans, 20);
    lv_span_set_mode(spans, LV_SPAN_MODE_BREAK);

    lv_span_t * span = lv_span_create(spans);
    lv_span_set_text(span, "china is a beautiful country.");
    lv_style_set_text_color(&span->style, lv_palette_main(LV_PALETTE_RED));
    lv_style_set_text_decor(&span->style, LV_TEXT_DECOR_STRIKETHROUGH | LV_TEXT_DECOR_UNDERLINE);
    lv_style_set_text_opa(&span->style, LV_OPA_30);

    span = lv_span_create(spans);
    lv_span_set_text_static(span, "good good study, day day up.");
    lv_style_set_text_font(&span->style,  &lv_font_montserrat_24);
    lv_style_set_text_color(&span->style, lv_palette_main(LV_PALETTE_GREEN));

    span = lv_span_create(spans);
    lv_span_set_text_static(span, "LVGL is an open-source graphics library");
    lv_style_set_text_color(&span->style, lv_palette_main(LV_PALETTE_BLUE));

    span = lv_span_create(spans);
    lv_span_set_text_static(span, "the boy no name.");
    lv_style_set_text_color(&span->style, lv_palette_main(LV_PALETTE_GREEN));
    lv_style_set_text_font(&span->style, &lv_font_montserrat_22);
    lv_style_set_text_decor(&span->style, LV_TEXT_DECOR_UNDERLINE);

    span = lv_span_create(spans);
    lv_span_set_text(span, "I have a dream that hope to come true");

    lv_span_refr_mode(spans);

    //lv_span_del(spans, span);
    //lv_obj_del(spans);
}
```
image

File: src/extra/widgets/span/lv_span.h [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+ *      TYPEDEFS
+ **********************/
+
+enum {
+    LV_SPAN_ALIGN_LEFT,    /**< Align text to left */
+    LV_SPAN_ALIGN_CENTER,  /**< Align text to center */
+    LV_SPAN_ALIGN_RIGHT,   /**< Align text to right */
+    LV_SPAN_ALIGN_AUTO,    /**< Use LEFT or RIGHT depending on the direction of the text (LTR/RTL)*/
+};
+typedef uint8_t lv_span_align_t;
```
Comment (kisvegabor) [inline src/extra/widgets/span/lv_span.h]: Can we use `lv_text_align_t` instead? 
In v8 it's style property (called `text_align`). I think it's reasonable to follow this pattern here too.

File: src/extra/widgets/span/lv_span.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+ * get max line height of all span in the spangroup.
+ * @param obj pointer to a spangroup object.
+ */
+lv_coord_t lv_span_get_max_line_h(lv_obj_t * obj);
+
+/**
+ * get the width when all span of spangroup on a line. include spangroup pad.
+ * @param obj pointer to a spangroup object.
+ */
+lv_coord_t lv_span_get_expend_width(lv_obj_t * obj);
```
Comment (kisvegabor) [inline src/extra/widgets/span/lv_span.h]: `expend` -> `expand`

File: src/extra/widgets/span/lv_span.h [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+/**
+ * get the height with width fixed. the height include spangroup pad.
+ * @param obj pointer to a spangroup object.
+ */
+lv_coord_t lv_span_get_expend_height(lv_obj_t * obj, lv_coord_t width);
```
Comment (kisvegabor) [inline src/extra/widgets/span/lv_span.h]: `expend` -> `expand`

File: src/extra/widgets/span/lv_span.c [hunk 4]
```diff
@@ -0,0 +1,966 @@
...
+    return ret;
+}
+
+/**
+ * draw span group
+ * @param spans obj handle
+ * @param coords coordinates of the label
+ * @param mask the label will be drawn only in this area
+ */
+static LV_ATTRIBUTE_FAST_MEM void lv_draw_span(lv_obj_t * obj, const lv_area_t * coords, const lv_area_t * mask)
```
Comment (kisvegabor) [inline src/extra/widgets/span/lv_span.c]: I think this function is not that performance-critical to be in `FAT_MEM`.

**Review by kisvegabor — COMMENTED** [review]
Amazing,  thank you very much for poting to v8.

I like your implementation. It's clear and well organized.

Besides the minor comments above I have some smaller ideas, but let's do smaller steps.

We still need 2 things to make this widget easy to use for the users:
- documentation: a markdown file similar to [bar.md](https://github.com/lvgl/lvgl/blob/master/docs/widgets/bar.md).
- some examples [here](https://github.com/lvgl/lvgl/tree/master/examples/widgets). The example in your comment would be perfect.

@turoksama We have another great new widget. Could you also take look at it?

File: src/extra/widgets/span/lv_span.h (hunk 3 above)
Comment (turoksama) [reply src/extra/widgets/span/lv_span.h]: lv_span_get_expend_height
may mean
lv_span_get_height directly？

**Comment by turoksama** [comment]
> Amazing, thank you very much for poting to v8.
> 
> I like your implementation. It's clear and well organized.
> 
> Besides the minor comments above I have some smaller ideas, but let's do smaller steps.
> 
> We still need 2 things to make this widget easy to use for the users:
> 
> * documentation: a markdown file similar to [bar.md](https://github.com/lvgl/lvgl/blob/master/docs/widgets/bar.md).
> * some examples [here](https://github.com/lvgl/lvgl/tree/master/examples/widgets). The example in your comment would be perfect.
> 
> @turoksama We have another great new widget. Could you also take look at it?

Yes,  noticed too.  I will test it very soon.

**Comment by guoweilkd** [comment]
The build failed,  need to open some fonts macro. which fonts are available on current environment?

**Comment by kisvegabor** [comment]
Only `LV_FONT_DEFAULT` exists for sure. I suggest adding macros to set different fonts. E.g. 
```c
#if LV_FONT_MONTSERRAT_22
    lv_style_set_text_font(&span->style, &lv_font_montserrat_22);
#endif
```

**Comment by kisvegabor** [comment]
Thanks! 
The Micropython test fails due to this:
```
../lib/lv_bindings/lvgl/src/extra/widgets/span/lv_span.c:10:10: fatal error: ../draw/lv_draw.h: No such file or directory
   10 | #include "../draw/lv_draw.h"
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```

You can simply include `lvgl.h` like here: https://github.com/lvgl/lvgl/blob/master/src/extra/widgets/chart/lv_chart.h#L16

**Review by turoksama — APPROVED** [review]
@guoweilkd Nice! It seems fine to me.
@kisvegabor

**Review by embeddedt — APPROVED** [review]
Looks good to me.

An idea: the implementation could later be made into a proper inline layout in v9 (which would allow us to drop the spangroup widget, and also allow features like inline images in text).

**Comment by kisvegabor** [comment]
Cool, I merge it. :slightly_smiling_face: 

@embeddedt 
> An idea: the implementation could later be made into a proper inline layout in v9 (which would allow us to drop the spangroup widget, and also allow features like inline images in text).

A proper inline layout would be really great. It'd be unique to LVGL among the embedded UI libraries (AFAIK at least)

**Comment by kisvegabor** [comment]
@guoweilkd 
Could you take a look at this comment?
https://github.com/lvgl/lvgl/issues/1763#issuecomment-841513375

**Comment by sakumisu** [comment]
Please change your span code in here,do not use STRIKETHROUGH.
`    lv_span_set_text(span, "china is a beautiful country.");
    lv_style_set_text_color(&span->style, lv_palette_main(LV_PALETTE_RED));
    lv_style_set_text_decor(&span->style, LV_TEXT_DECOR_STRIKETHROUGH | LV_TEXT_DECOR_UNDERLINE);`

this  will bring ambiguity for chinese people. Thank you!

**Comment by kisvegabor** [comment]
Thanks for pointing it out. I've fixed it here: 7bb09e35
The docs will be rebuilt in ~10 minutes.

**Pull request merged**